### PR TITLE
Add support for the LoongArch architecture.

### PR DESCRIPTION
**Pull request by yetist**
LoongArch is a new instruction set of Loongson 3A5000 CPU, you can read the documents or visit the development community to get more infomation.

This PR add support for the LoongArch architecture, and the related PR is https://github.com/systemd/systemd/pull/20087.

### CODE REVIEW

File: src/basic/architecture.h [hunk 1]
```diff
@@ -229,6 +230,9 @@ int uname_architecture(void);
 #    define native_architecture() ARCHITECTURE_ARC
 #    define LIB_ARCH_TUPLE "arc-linux"
 #  endif
+#elif defined(__loongarch64)
+#    define native_architecture() ARCHITECTURE_LOONGARCH64
+#    define LIB_ARCH_TUPLE "loongarch64-linux-gnu"
```
Comment (xen0n) [inline src/basic/architecture.h]: This might need revision due to ongoing discussion at loongson/LoongArch-Documentation#25?

**Review by xen0n — COMMENTED** [review]
The commit message of first commit is inconsistent with the 2nd commit in style; please unify them.

**Review by yuwata — APPROVED** [review]
Not familiar with the architecture. But superficially LGTM. At least, this should not break any existing architectures.

File: src/basic/architecture.h (hunk 1 above)
Comment (keszybz) [reply src/basic/architecture.h]: Hmm, I don't see anything in that pull request diff that'd influence this triplet (and I can't read the discussion). What part do you have in mind?

Comment (xen0n) [reply src/basic/architecture.h]: I mean the part touching the "multiarch tuple" section; in the latest revision of ABI spec and toolchain code, the ABI is to be explicitly marked in the tuples, like:

- `loongarch64-linux-gnuf64` for the LP64D ABI (basically means the same as RISC-V LP64D),
- `loongarch64-linux-gnuf32` for the LP64F ABI (ditto),
- `loongarch64-linux-gnusf` for the LP64S ABI (the soft-float ABI is marked with "S" in LoongArch).

Therefore, this tuple should be updated as well when the spec update gets merged; we might want to wait for gcc, for definitive values to use here, to reduce code churn.

Comment (keszybz) [reply src/basic/architecture.h]: Yeah, I guess we should wait then.

**Comment by poettering** [comment]
BTW, this newly added documentation might be interesting here:

https://systemd.io/PORTING_TO_NEW_ARCHITECTURES

**Comment by DaanDeMeyer** [comment]
@xen0n @yetist What's the status of this? This is blocking #21487 because it needs the architecture to be defined for loongarch. If this will still take a while maybe we should drop the gpt commit for loongarch and re-introduce it with this PR

**Comment by poettering** [comment]
maybe leave `LIB_ARCH_TUPLE` undefined for now for the arch, and then add some minimal `#ifdef` for the few users for that (and maybe add a `#pragma message` line to remind maintainers to add this eventually). That way we can sidestep the issue of the final LIB_ARCH_TUPLE for now.

**Comment by yetist** [comment]
> maybe leave `LIB_ARCH_TUPLE` undefined for now for the arch, and then add some minimal `#ifdef` for the few users for that (and maybe add a `#pragma message` line to remind maintainers to add this eventually). That way we can sidestep the issue of the final LIB_ARCH_TUPLE for now.

I updated this branch and added `#pragma message` to remind me to make changes in the future.

File: src/basic/architecture.h [hunk 2]
```diff
@@ -43,6 +43,7 @@ enum {
         ARCHITECTURE_TILEGX,
         ARCHITECTURE_X86,
         ARCHITECTURE_X86_64,
+        ARCHITECTURE_LOONGARCH64,
```
Comment (poettering) [inline src/basic/architecture.h]: this was recently sorted properly, please keep this sorted

**Comment by yetist** [comment]
@DaanDeMeyer @xen0n @poettering 

Because this PR blocking #21487, so I think we can merge this PR first and submit all other [loongarch patches](https://github.com/loongarch64/systemd/commits/dev-wu-main) then.

About the Tuple issue, after the psABI document is stable, I will submit it again.

What do you think?

File: src/basic/architecture.h (hunk 2 above)
Comment (yetist) [reply src/basic/architecture.h]: Done.

**Comment by keszybz** [comment]
```
Assertion 'gpt_partition_type_uuid_to_arch(id) == a' failed at src/test/test-gpt.c:49, function test_gpt_types_against_architectures(). Aborting.
```

**Comment by keszybz** [comment]
We added a bunch of tests for the consistency of those definitions, please make sure to run tests locally.

File: src/basic/architecture.c [hunk 3]
```diff
@@ -66,6 +66,9 @@ int uname_architecture(void) {
 #elif defined(__ia64__)
                 { "ia64",       ARCHITECTURE_IA64     },
 
+#elif defined(__loongarch64)
+                { "loongarch64", ARCHITECTURE_LOONGARCH64 },
+
```
Comment (keszybz) [inline src/basic/architecture.c]: `l` is before `m`. Please move this before `m68k`.

Also, please drop the dot from the commit message :pray:

File: src/basic/architecture.c [hunk 4]
```diff
@@ -148,6 +151,7 @@ static const char *const architecture_table[_ARCHITECTURE_MAX] = {
         [ARCHITECTURE_X86_64]    = "x86-64",
         [ARCHITECTURE_X86]       = "x86",
         [ARCHITECTURE_IA64]      = "ia64",
+        [ARCHITECTURE_LOONGARCH64] = "loongarch64",
```
Comment (keszybz) [inline src/basic/architecture.c:154]: Please reindent the whole table to match again.

File: src/basic/architecture.h [hunk 5]
```diff
@@ -227,6 +228,10 @@ int uname_architecture(void);
 #    define native_architecture() ARCHITECTURE_ARC
 #    define LIB_ARCH_TUPLE "arc-linux"
 #  endif
+#elif defined(__loongarch64)
```
Comment (keszybz) [inline src/basic/architecture.h]: Hmm, the order here is a mess. But please insert the new addition here right above `m68k` too.

**Review by keszybz — CHANGES_REQUESTED** [review]

**Comment by yetist** [comment]
The tests results:
```
Ok:                 933 
Expected Fail:      0   
Fail:               0   
Unexpected Pass:    0   
Skipped:            20  
Timeout:            0   

```

File: src/basic/architecture.c (hunk 3 above)
Comment (yetist) [reply src/basic/architecture.c]: Done.

**Pull request merged**